okhttp/src/main/java/okhttp3/internal/cache/CacheInterceptor.java

```diff
@@ -101,8 +101,16 @@ public CacheInterceptor(InternalCache cache) {
     // If we have a cache response too, then we're doing a conditional get.
     if (cacheResponse != null) {
       if (networkResponse.code() == HTTP_NOT_MODIFIED) {
+        Headers.Builder newResponseHeadersBuilder = combine(cacheResponse.headers(), networkResponse.headers()).newBuilder();
+
+        // If we have Content-Encoding IN the new response, but NOT in the old response our response body will be broken
+        // that's why we ignore the new Content-Encoding header. See https://github.com/square/okhttp/pull/3700 for more information.
+        if (!cacheResponse.headers().names().contains("Content-Encoding")) {
```

Comment:
names requires building a treeset, I think checking on headers() directly would be quicker.  Also it is probably better to unconditionally remove as even a change to Content-Encoding would break things.

Reply:
@yschimke good point!

I changed it to *remove* Content-Encoding from combined headers, if it was not present in the cached headers. And keep the cached header if it was present in the cached headers.